src/wh_client_crypto.c

```diff
@@ -0,0 +1,697 @@
+                evict = 0;
+
+                /* Response Message */
+                wh_Packet_pk_ecc_sign_res* res;
+                res = &packet->pkEccSignRes;
+                uint8_t* res_out = (uint8_t*)(res + 1);
+                uint16_t res_len;
+
+                /* Recv Response */
+                do {
+                    ret = wh_Client_RecvResponse(ctx, &group, &action, &res_len,
+                        (uint8_t*)packet);
+                } while (ret == WH_ERROR_NOTREADY);
+
+                if (ret == WH_ERROR_OK) {
+                    if (packet->rc == 0) {
+                        uint16_t out_len = res->sz;
+                        /* check inoutlen and read out */
+                        if (inout_len != NULL) {
+                            if (out_len > *inout_len) {
+                                /* Silently truncate the signature */
+                                out_len = *inout_len;
+                            }
+                            *inout_len = out_len;
+                            if (    (out != NULL) &&
+                                    (out_len > 0)) {
+                                memcpy(out, res_out, out_len);
+                            }
+                        }
+                    } else {
+                        ret = packet->rc;
+                    }
+                }
+            }
+        } else {
+            /* Request length is too long */
+            ret = WH_ERROR_BADARGS;
+        }
+    }
+    /* Evict the key manually on error */
+    if(evict != 0) {
+        (void)wh_Client_KeyEvict(ctx, key_id);
+    }
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+    printf("[client] %s ret:%d\n", __func__, ret);
+#endif
+    return ret;
+}
+
+int wh_Client_EccDsaVerify(whClientContext* ctx, ecc_key* key,
+        const uint8_t* sig, word32 sig_len,
+        const uint8_t* hash, word32 hash_len,
+        int *out_res)
+{
+    int ret = 0;
+    whPacket* packet;
+
+    /* Transaction state */
+    whKeyId key_id;
+    int evict = 0;
+    int export_pub_key = 0;
+
+
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+printf("[client] %s ctx:%p key:%p, in:%p in_len:%u, out:%p inout_len:%p\n",
+        __func__, ctx, key, in, in_len, out, inout_len);
+#endif
+
+    if (    (ctx == NULL) ||
+            (key == NULL) ||
+            ((sig == NULL) && (sig_len > 0)) ||
+            ((hash == NULL) && (hash_len > 0)) ) {
+        return WH_ERROR_BADARGS;
+    }
+
+    packet = (whPacket*)wh_CommClient_GetDataPtr(ctx->comm);
+    if (packet == NULL) {
+        return WH_ERROR_BADARGS;
+    }
+    memset((uint8_t*)packet, 0, sizeof(*packet));
+
+    /* TODO: Check the request size to ensure it will fit before importing key*/
+#if 0
+    if (req_len > WOLFHSM_CFG_COMM_DATA_LEN) {
+        ret = WH_ERROR_BADARGS;
+        break;
+    }
+#endif
+
+    key_id = WH_DEVCTX_TO_KEYID(key->devCtx);
+    if (key->type == ECC_PRIVATEKEY_ONLY) {
+            export_pub_key = 1;
+        }
```

Comment:
There is a test in wolfCrypt that invokes verify using a private only key.  wolfCrypt will normally invoke make_pub and update the ecc_key with the now computed public key and then invoke verify.  Afterward, the test then checks to make sure the public key was updated in the ecc_key.  So, we have to ask the server to send the public key afterward so the local version of the ecc_key will have the public key set.  The expected use case here is the client has only a private key id (it's really stored on the server), and wants to the server to verify a signature.  The server does that and returns the public key as well, which we dutifully update the local key structure with.  We needed to avoid invoking make_pub locally since that is not routed through cryptocb.  Not exactly a huge fan of this dance, but it does save a few round trips.